src/cj.c

```diff
@@ -0,0 +1,66 @@
+#include "data.table.h"
+
+SEXP cj(SEXP base_list) {
+  int JJ = LENGTH(base_list), nprotect = 0;
+  SEXP out = PROTECT(allocVector(VECSXP, JJ)); nprotect++;
+  int NN = 1;
+  for (int j = 0; j < JJ; j++) NN *= LENGTH(VECTOR_ELT(base_list, j));
+  int div = NN, modulo;
+  
+  for (int j = 0; j < JJ; j++) {
+    SEXP this_v = VECTOR_ELT(base_list, j);
+    modulo = div;
+    div = modulo/LENGTH(VECTOR_ELT(base_list, j));
+    switch(TYPEOF(this_v)) {
+    case LGLSXP: {
+      SEXP this_col = PROTECT(allocVector(LGLSXP, NN)); nprotect++;
+      int *this_v_ptr = LOGICAL(this_v);
+      int *this_col_ptr = LOGICAL(this_col);
+      for (int i = 0; i < NN; i++) {
+        this_col_ptr[i] = this_v_ptr[(i % modulo) / div];
+      }
+      SET_VECTOR_ELT(out, j, this_col);
+    }
+      break;
+    case INTSXP: {
+      SEXP this_col = PROTECT(allocVector(INTSXP, NN)); nprotect++;
+      int *this_v_ptr = INTEGER(this_v);
+      int *this_col_ptr = INTEGER(this_col);
```

Comment:
@mattdowle is this the proper way? (also for `VECSXP`/`STRSXP` cases it seems there's no analog?)

Reply:
Correct.  You can't do it for VECSXP/STRSXP because that would be what's known as "behind the write barrier".  You can partially do it though by reading (RHS) but not writing (LHS =).

Reply:
I'm off to bed now an it's a long weekend here -- so have fun in C!   Where did you get the sense that CJ() was slow and that it could be sped up -- is the benchmark you put in this PR what motivated you or is there another one?
The idea from way-back was that CJ() would not actually materialize the full table.  It would just mark itself as "CJ" object and then bmerge would know how to recycle from the irregular non-expanded structure appropriately without needing to expand it.

Reply:
No benchmark, just from staring at the code now and then and thinking "there's gotta be a faster way", and then happening on the modulo/div version here in my scratch work on an unrelated issue but alas. I'm still surprised how fast it is... `rep.int` is a beast. I guess `rep.int` is doing most of the work, and the recursion is only touching scalars instead of the full data (so it's not slowing down).

The alt-rep-y form does sound promising (certainly way more memory efficient), though I do use `CJ` often outside of `i` for populating the crossjoin tables for other reasons.

Reply:
enjoy the weekend!

Reply:
yep: CJ uses rep.int and its implemented in C and a .Internal() R function too (which makes it fast to call in a loop).  So really your're trying to beat C with C.  `Rprof` on the current CJ might be misleading actually as I think someone said `Rprof` doesn't count .Internal() R functions.
Are definitely getting the same result in the same order?  Doing "(i % modulo) / div" on every single iteration is inefficient because it's really just batches of increasing i.  It's possible that all those % and / add up in this case (it's contiguous assign so cache effects shouldn't come into play).

Reply:
> it's really just batches of increasing i

Not sure I follow this.

Latest commit splits out the first and last column as separate cases to skip one of the `%`/`/` operations, gave about 10% boost.

Reply:
generally better to parallelise upper layers if possible, thus parallelising columns should give better improvement